core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -0,0 +1,180 @@
+      // reconnected so return the retriable network exception.
+      if (currentTransactionData != null) {
+        val error = if (currentTransactionData.producerEpoch() < transactionData.producerEpoch())
+          Errors.INVALID_PRODUCER_EPOCH
+        else 
+          Errors.NETWORK_EXCEPTION
+        val topicPartitionsToError = mutable.Map[TopicPartition, Errors]()
+        currentTransactionData.topics().forEach { topic =>
+          topic.partitions().forEach { partition =>
+            topicPartitionsToError.put(new TopicPartition(topic.name(), partition), error)
+          }
+        }
+        val oldCallback = currentNodeAndTransactionData.callbacks(transactionData.transactionalId())
+        currentNodeAndTransactionData.transactionData.remove(transactionData)
+        oldCallback(topicPartitionsToError.toMap)
+      }
+      currentNodeAndTransactionData.transactionData.add(transactionData)
+      currentNodeAndTransactionData.callbacks.put(transactionData.transactionalId(), callback)
+      wakeup()
+    }
+  }
+
+  private class AddPartitionsToTxnHandler(node: Node, transactionDataAndCallbacks: TransactionDataAndCallbacks) extends RequestCompletionHandler {
+    override def onComplete(response: ClientResponse): Unit = {
+      // Note: Synchronization is not needed on inflightNodes since it is always accessed from this thread.
+      inflightNodes.remove(node)
+      if (response.authenticationException() != null) {
+        error(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with an " +
+          "authentication exception.", response.authenticationException)
+        transactionDataAndCallbacks.callbacks.foreach { case (txnId, callback) =>
+          callback(buildErrorMap(txnId, transactionDataAndCallbacks.transactionData, Errors.forException(response.authenticationException()).code()))
+        }
+      } else if (response.versionMismatch != null) {
+        // We may see unsupported version exception if we try to send a verify only request to a broker that can't handle it. 
+        // In this case, skip verification.
+        error(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with invalid version exception. This suggests verification is not supported." +
+              s"Continuing handling the produce request.")
+        transactionDataAndCallbacks.callbacks.values.foreach(_(Map.empty))
+      } else {
+        val addPartitionsToTxnResponseData = response.responseBody.asInstanceOf[AddPartitionsToTxnResponse].data
+        if (addPartitionsToTxnResponseData.errorCode != 0) {
+          error(s"AddPartitionsToTxnRequest for broker ${config.brokerId}  returned with error ${Errors.forCode(addPartitionsToTxnResponseData.errorCode)}.")
+          // The client should not be exposed to CLUSTER_AUTHORIZATION_FAILED so modify the error to invalid record -- to signify the verification did not complete.
+          // Older clients return with INVALID_RECORD
```

Comment:
Hmm, not quite sure that I follow. Do you mean that we return INVALID_RECORD to be compatible for old clients?

Reply:
Yes. Sorry for the confusion. In the KIP we mention INVALID_TXN_STATE for the new clients, but old clients use INVALID_RECORD for compatibility. I will update the comment.